lib/eosknowledgeprivate/ekn-media-bin.c

```diff
@@ -0,0 +1,1577 @@
+  gint i, n;
+
+  for (i = 0, n = gst_tag_list_get_tag_size (list, tag); i < n; ++i)
+    {
+      const GValue *val = gst_tag_list_get_value_index (list, tag, i);
+      GValue str = {0, };
+
+      g_value_init (&str, G_TYPE_STRING);
+      g_value_transform (val, &str);
+
+      g_string_append_printf (metadata->tag, "\n    %s", tag);
+      g_string_append_printf (metadata->val, "\n: %s", g_value_get_string (&str));
+
+      g_value_unset (&str);
+    }
+}
+
+static inline void
+meta_data_strings_set_info (MetaDataStrings *metadata,
+                            GtkLabel        *left,
+                            GtkLabel        *right,
+                            GstTagList      *tags,
+                            const gchar     *title)
+{
+  if (tags)
+    {
+      g_string_assign (metadata->tag, title);
+      g_string_assign (metadata->val, "");
+
+      gst_tag_list_foreach (tags, print_tag, metadata);
+
+      gtk_label_set_label (left, metadata->tag->str);
+      gtk_label_set_label (right, metadata->val->str);
+    }
+  else
+    {
+      gtk_label_set_label (left, "");
+      gtk_label_set_label (right, "");
+    }
+}
+
+static inline void
+ekn_media_bin_update_stream_info (EknMediaBin *self)
+{
+  MetaDataStrings metadata = { g_string_new (""), g_string_new ("") };
+
+  meta_data_strings_set_info (&metadata,
+                              self->info_1_label,
+                              self->info_2_label,
+                              self->audio_tags,
+                              "Audio:");
+
+  meta_data_strings_set_info (&metadata,
+                              self->info_3_label,
+                              self->info_4_label,
+                              self->video_tags,
+                              "Video:");
+
+  meta_data_strings_set_info (&metadata,
+                              self->info_5_label,
+                              self->info_6_label,
+                              self->text_tags,
+                              "Text:");
+
+  g_string_free (metadata.tag, TRUE);
+  g_string_free (metadata.val, TRUE);
+}
+
+static inline void
+ekn_media_bin_handle_msg_application (EknMediaBin *self, GstMessage *msg)
+{
+  const GstStructure *structure;
+  const gchar *name;
+
+  structure = gst_message_get_structure (msg);
+  name = gst_structure_get_name (structure);
+  g_return_if_fail (name != NULL);
+
+  GST_DEBUG ("Application message: %" GST_PTR_FORMAT, structure);
+
+  if (self->show_stream_info)
+    ekn_media_bin_update_stream_info (self);
+
+  /* TODO: handle audio and text tags */
+  if (g_str_equal (name, "video-tags-changed"))
+    {
+      gchar *value = NULL;
+
+      if (!self->title_user_set)
+        {
+          if (self->video_tags)
+            gst_tag_list_get_string_index (self->video_tags, GST_TAG_TITLE, 0, &value)
+
+          ekn_media_bin_set_title (self, (self->video_tags && value) ? value : "");
+          self->title_user_set = FALSE;
+          g_free (value);
```

Comment:
nitpick: You could use `g_clear_pointer(&value, g_free)` here to save the extra assignment of `NULL` in the case where description is not set and title is